JK: add BluetoothAdvertisingDevices

J0EK3R opened the pull request:
Here is a first try for a management of - I called them - StaticDevices with an editor to select...
The name "StaticDevices" isn't that good... suggestions?

Commit:
JK: added BluetoothAdvertisingDevice classes and device MK6

Commit:
JK: added ressource text for OpenDeviceInsertDialog

Commit:
JK: fix - wrong devicetype

Commit:
JK: changed button order

Commit:
JK: added StaticDeviceManager

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,76 @@
+#pragma warning disable CS8600 // Converting null literal or possible null value to non-nullable type.
+#pragma warning disable CA1416 // Validate platform compatibility
+
+namespace BrickController2.Droid.PlatformServices.BluetoothLE
```
Nit pick: I personally prefer file scope namepsaces as it saves spacing.

J0EK3R replied:
Check:
https://github.com/J0EK3R/brickcontroller2/commit/36d11db608f05b355bddb0a4c51538c91eb1b070

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,76 @@
+                null,
+                this);
+
+            return await Task.FromResult(true);
```
Nit pick: There is no need to await this if you remove `async` keyword for this method.

vicocz replied:
Seems this dhould be applied several times....

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,76 @@
+        public async Task<bool> StopAdvertiseAsync()
+        {
+            _advertiser?.StopAdvertisingSet(this);
+
```
I miss reset of `_advertisingSet`

J0EK3R replied:
Here we go:
https://github.com/J0EK3R/brickcontroller2/commit/3054fc94fd8e5feb8d2c78ad693c32a49a0ed725

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,76 @@
+            return await Task.FromResult(true);
+        }
+
+        public bool ChangeAdvertiseAsync(ushort manufacturerId, byte[] rawData)
```
No `Async` ending as this is not asynchronous.
Nit pick: may be `UpdateAdvetisedData` or something similar will explain it more

J0EK3R replied:
I reworked that:
https://github.com/J0EK3R/brickcontroller2/commit/05c9349369890c09155e3058734a7a4e9e1874eb

vicocz commented on BrickController2/BrickController2/BrickController2.csproj:
```diff
@@ -50,4 +52,18 @@
   </ItemGroup>
 
+  <ItemGroup>
+    <Compile Update="UI\Pages\StaticDeviceListPage.xaml.cs">
```
This should not be needed.

J0EK3R replied:
Yes, you're right:
https://github.com/J0EK3R/brickcontroller2/commit/c12e89bc83d9e89ebc2d98b7a1badf39d66514d7

J0EK3R replied:
Puuh - I was completely wrong - I had to revert that commit:

https://github.com/J0EK3R/brickcontroller2/commit/13b261f5c410b5f0b83a62cff460c8565e410573

And here is the cleaned-up project file:
https://github.com/J0EK3R/brickcontroller2/commit/69f902d3e91f30e6c7b4932f0d957ffb77af1e63

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+        /// <summary>
+        /// timespan after TryGetTelegram is called from output loop to refresh data
+        /// </summary>
+        private TimeSpan _cyclicDataRefreshTimeSpan = TimeSpan.FromSeconds(2);
```
Nit pick: missing `readonly`.

J0EK3R replied:
Fixed:
https://github.com/J0EK3R/brickcontroller2/commit/36d526b55510d126bc98ac2a17638956044dd7c9

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+        /// <summary>
+        /// timespan to wait after each output loop
+        /// </summary>
+        private TimeSpan _cyclicLoopWaitTimeSpan = TimeSpan.FromMilliseconds(100);
```
Nit pick: missing `readonly`.

J0EK3R replied:
OK:
https://github.com/J0EK3R/brickcontroller2/commit/36d526b55510d126bc98ac2a17638956044dd7c9

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+                {
+                    await DisconnectInternalAsync();
+
+                    Log.Error($"Connection exception: {exception.Message}");
```
I suggest to use Microsoft.Extensions.Logging's `ILogger` passed via constructor.

J0EK3R replied:
I completely removed the logging...:
https://github.com/J0EK3R/brickcontroller2/commit/2620b908ff547b7fbd869f5b037b9e34a88d975b

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+        #endregion
+        #region DisconnectAsync
+        /// <summary>
+        /// stop output loop and dispos the advertising device
```
Nitpick: typo `dispos`

J0EK3R replied:
Fixed:
https://github.com/J0EK3R/brickcontroller2/commit/36d526b55510d126bc98ac2a17638956044dd7c9

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+        /// <summary>
+        /// stop output loop and dispos the advertising device
+        /// </summary>
+        /// <returns>Task</returns>
```
Does it make sense to have it here if there is no description. I guess Task is the default type returned by asyncronous method.

J0EK3R replied:
Fixed:
https://github.com/J0EK3R/brickcontroller2/commit/36d526b55510d126bc98ac2a17638956044dd7c9

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+                {
+                    byte[] currentData;
+                    if (_bleAdvertiserDevice != null &&
+                        TryGetTelegram(out currentData))
```
Nit pick: `TryGetTelegram(out var currentData))` to save its definition.

J0EK3R replied:
Done:
https://github.com/J0EK3R/brickcontroller2/commit/ffc6937fc1c6c5d5574b4b5e0d4a3d066bfcb7c1

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+                        stopwatch.Elapsed > _cyclicDataRefreshTimeSpan)
+                    {
+                        byte[] currentData;
+                        if (TryGetTelegram(out currentData))
```
`out var currentData` please.

J0EK3R replied:
Done:
https://github.com/J0EK3R/brickcontroller2/commit/ffc6937fc1c6c5d5574b4b5e0d4a3d066bfcb7c1

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+        /// <summary>
+        /// set device to initial state before output loop starts
+        /// </summary>
+        protected abstract void InitOutputTask();
```
DOas it make sense to have method with suffinx Task but doing nothing asynchronous?
I would either:
- made id asynchronous
- rename it

J0EK3R replied:
I renamed the method:
https://github.com/J0EK3R/brickcontroller2/commit/79f8a9074052594e41f2c1e597791507121a8dd5

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDevice.cs:
```diff
@@ -0,0 +1,317 @@
+        /// </summary>
+        /// <param name="currentData">ref to byte array</param>
+        /// <returns>True: success. False: no success</returns>
+        public abstract bool TryGetTelegram(out byte[] currentData);
```
Why is this `public` if it's called from `ProcessOutputAsync`? `protected` seems reasonable.

J0EK3R replied:
I changed that:
https://github.com/J0EK3R/brickcontroller2/commit/3c1fd13ceb4984f42bf2723c310a476c7a8be8df

vicocz commented on BrickController2/BrickController2/DeviceManagement/DI/DeviceManagementModule.cs:
```diff
@@ -11,6 +12,7 @@ protected override void Load(ContainerBuilder builder)
 
             builder.RegisterType<DeviceRepository>().As<IDeviceRepository>().SingleInstance();
             builder.RegisterType<DeviceManager>().As<IDeviceManager>().SingleInstance();
+            builder.RegisterType<StaticDeviceManager>().As<IStaticDeviceManager>().SingleInstance();
```
StaticDevice sounds badly, will think about naming.

J0EK3R replied:
Yes, all my ideas were bad:
* nonScannableDevice ...grrrr
* manualDevice ...hmmm

These devices aren't scannable and have to be inserted manually by selection...

In future there could be others like...
* mqttDevice
* udpDevice
* httpDevice
...whatelse is controlable by a gamepad! ;)

vicocz replied:
I have no better name right now, so let's keep this one. It can be changed in future if there is any better....

J0EK3R replied:
OK - ManualDevice was a bit better name - but let's rename the thing in future.

vicocz commented on BrickController2/BrickController2/DeviceManagement/MK6.cs:
```diff
@@ -0,0 +1,75 @@
+    /// <summary>
+    /// MK 6.0 Module
+    /// </summary>
+    internal class MK6 : MKBaseByte
```
Could we have per vendor folders please?
Add all MK e.g. to `MouldKing` folder

J0EK3R replied:
Yes, I would welcome that.
Let's put all devices (Lego, Buwizz, SBrick, aso.) in vendor folders!

I don't know the legal stuff about using the vendor's names - how should we continue - use the vendor name or not?

J0EK3R replied:
And what about the pictures – do we take them from somewhere on the internet or do we take our own photos?

vicocz replied:
> Yes, I would welcome that. Let's put all devices (Lego, Buwizz, SBrick, aso.) in vendor folders!
> 
> I don't know the legal stuff about using the vendor's names - how should we continue - use the vendor name or not?

Putting all other devices into separate folder is long term goal, not necessary for MK part.
I would like to keep this fork in reasonable diff state against the original repo.

J0EK3R replied:
OK, I'll move the MK stuff in a folder.

But we have to clarify the naming - is it legal to use the vendors names?

...how should we continue - use the vendor name or not?

Lego, Buwizz and the Others are used...

vicocz replied:
I guess no one will object related to naming folder based on vendor's name.
So please use vendor folder for your newly added MK* classes.

J0EK3R replied:
Yepp, git-moved them...

https://github.com/J0EK3R/brickcontroller2/commit/16144b06cfa21d7bd075767ee14b8ed492e62e74

vicocz commented on BrickController2/BrickController2/DeviceManagement/MK6.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <returns>Base-Telegram</returns>
+        private static byte[] GetTelegramBase(string address)
+        {
+            switch (address)
```
Nit pick: use pattern matching please:
```
            return address switch
            {
                MK6.Device3 => Telegram_Base_Device_3,
                MK6.Device2 => Telegram_Base_Device_2,
                _ => Telegram_Base_Device_1,
            };
```

J0EK3R replied:
Oh, I did not know that - changed:
https://github.com/J0EK3R/brickcontroller2/commit/3b8436a6ab254f2027570824eff71b88d4fd62f1

vicocz commented on BrickController2/BrickController2/DeviceManagement/MK6.cs:
```diff
@@ -0,0 +1,75 @@
+                case MK6.Device2:
+                    return Telegram_Base_Device_2;
+                case MK6.Device1:
+                default:
```
`default` should have own handling and thorw an exception. To see if there is wrong input.

J0EK3R replied:
unhandled pattern - default - throws exception now:
https://github.com/J0EK3R/brickcontroller2/commit/3b8436a6ab254f2027570824eff71b88d4fd62f1

vicocz commented on BrickController2/BrickController2/DeviceManagement/MKBaseByte.cs:
```diff
@@ -0,0 +1,178 @@
+        }
+        #endregion
+
+        #region TryGetTelegram(out byte[] currentData)
```
Nitpick I dislike these per method regions.

J0EK3R replied:
Your repo, your rulez - I killed 'em all ;)
https://github.com/J0EK3R/brickcontroller2/commit/36d526b55510d126bc98ac2a17638956044dd7c9

vicocz commented on BrickController2/BrickController2/Protocols/MKProtocol.cs:
```diff
@@ -0,0 +1,101 @@
+/// <summary>
+/// static class wich implements the encryption algorithm for the advertising data
+/// </summary>
+public class MKProtocol
```
This should be `static` I guess

J0EK3R replied:
Done:
https://github.com/J0EK3R/brickcontroller2/commit/36d526b55510d126bc98ac2a17638956044dd7c9

vicocz reviewed (commented):
Left base comments, will continue (probably) during weekend with deeper review.
ANyway, this will be cool 😄

Commit:
JK: using of file scope namespaces

Commit:
JK: Bugfix for Android
https://github.com/dotnet/maui/issues/23358
removed unused stuff

Commit:
JK: removed unused stuff

Commit:
JK: reworked IBluetoothLEAdvertiserDevice
* changed async to sync
* renaming

Commit:
JK: reset _advertisingSet to null

Commit:
JK: rework
* removed all per-method-regions
* fixed typos
* removed description for async Task return value

Commit:
JK: removed logging

Commit:
JK: using inline definition

Commit:
JK: rename InitOutputTask -> InitDevice

Commit:
JK: changed TryGetTelegram to protected

Commit:
JK: use pattern matching

Commit:
JK: typo

J0EK3R commented:
I did a bugfix in StaticDeviceListPageViewModel - the generic Type of ObservableCollection SelectedFactoryDatas has to be `object` to work properly.
[Here is the issue](https://github.com/dotnet/maui/issues/23358).

But the selection dialog isn't that nice - any suggestions to pimp that? ;)

https://github.com/J0EK3R/brickcontroller2/commit/3749332f3348cf83b08b55ebe174526b3d0e8ef9

Commit:
Revert "JK: removed unused stuff"

<controls:Dialogs...> is needed for PageBase.FindDialogServer() not to throw NotImplementedException("No dialog server");

This reverts commit c12e89bc83d9e89ebc2d98b7a1badf39d66514d7.

Commit:
JK: removed not needed ItemGroups in Project

Commit:
JK: added BluetoothAdvertiser
to handle multiple BluetoothAdvertisingDevices sharing the same Bluetooth telegram (linke MK4)

Commit:
JK: catch possible exception when loading devices

Commit:
JK: some reworking

Commit:
JK: added MK4

J0EK3R commented:
I added a second MK device: MK4

The special about MK4 is that there is only one advertising telegram for three possible MK4 devices...

So I had to do some changes:
now there is a class `BluetoothAdvertiser `- MK6 devices have one BluetoothAdvertiser instance per device instance - MK4 devices have one BluetoothAdvertiser instance for all three possible device instances...

Commit:
JK: replaced property "DataVersion" by method "NotifyDataChanged"

vicocz commented on BrickController2/BrickController2/UI/Pages/DeviceListPage.xaml:
```diff
@@ -66,6 +66,7 @@
                     </CollectionView.EmptyView>
                 </CollectionView>
 
+                <controls:FloatingActionButton Style="{StaticResource FloatingActionButtonStyle}" Icon="playlist_add" Command="{Binding ShowStaticDeviceListPageCommand}" ToolTipProperties.Text="{extensions:Translate OpenStaticDeviceList}" HorizontalOptions="End" VerticalOptions="End" Margin="0,0,100,10" IsVisible="{Binding IsBLEAdvertisingSupported}"/>
```
To be honest, I've got completely lost when I pressed this button afor the first time. I even did not realized the new page was openned.

So I guess user scenario should be done somehow better.
How do you think user will add such MK device(s)? Does he want to select per vonder / type? DOes ho want to add multiple ones?

J0EK3R replied:
It's possible to have up to 3 MK4 and up to 3 MK6 - so the user should be able to add these...

Btw.: to change the device number on the physical device you short-press the on/off button.

J0EK3R replied:
It's possible to have up to 3 MK4 and up to 3 MK6 (with different addresses) - so the user should be able to add all of these...

Btw.: to change the device number on the physical device you short-press the on/off button.

J0EK3R replied:
Yes, you're right: I think the selection to add these non-scanable-devices should better be a Dialog than a page.
But the selection process should be like this.
Perhabs the already added devices could also be shown in the selection list - marked as added...

vicocz replied:
May be we need to add toolbar button per each device type and let user choose addresses via some selection dialog (that supports multiple selection). What do you think?

J0EK3R replied:
I think it's good like it is for each possible instance of those both MK devices.
An address bases selection makes no Sense for other devices to come: PowerBox (only one instance possible), MechanicakMaster (only one instance), other MK devices with only one possible instance...
And this list is growing...

vicocz replied:
Then my suggestion is:
- keep standalone dialog
- use different title than `Devices` as it's totally confusing
- use the different way how to display / select "static" devices
  - use grouped view
  - group label should
    - contain device type only - in case there are several addresses available
    - contains device type and name - in case there might be single instance only
    - there might be checkbox to select it
  - Back button does nothing
  - To add selected devices you need to press toolbar button as e.g. Device Settings has

![image](https://github.com/user-attachments/assets/61d7cbe4-c67e-4568-8e2d-a674f82b25f7)

J0EK3R replied:
Oh yeah, looks really good...

Would take some time - first I'll process your "nitpicks" ;)

J0EK3R replied:
Here we are:

![grafik](https://github.com/user-attachments/assets/0b807e41-7360-4cff-a882-b1b6e503088e)


https://github.com/J0EK3R/brickcontroller2/commit/38ddb07a95304c954631ec7d8e66a7e22b334a0a

vicocz replied:
Looks much better now. Nice.
Just styling:
- decrease spacing/padding for address rows as it takes a lot of space
- you should probably center address and the swich vertically

J0EK3R replied:
I tried to improve the styling

![grafik](https://github.com/user-attachments/assets/c526860f-7428-4716-b336-d83eb6af315f)


https://github.com/J0EK3R/brickcontroller2/commit/e8536d2868a5c5087bc37e8db95e89a5495fdb11

vicocz commented on BrickController2/BrickController2/BrickController2.csproj:
```diff
@@ -34,9 +38,9 @@
-	<MauiFont Include="Resources\Fonts\*" />
-  </ItemGroup>	
-	
+	  <MauiFont Include="Resources\Fonts\*" />
```
Nit pick, please keep indention via spaces.

J0EK3R replied:
Yes, fixed it and changed my VS setting to use spaces..

https://github.com/J0EK3R/brickcontroller2/commit/37d1a5f02fbdd4acf28d7df3941fad99133c20dd

...and fixed yours too...
![grafik](https://github.com/user-attachments/assets/b588205d-1556-4097-8f1b-eca18e84c71f)

vicocz commented on BrickController2/BrickController2/BrickController2.csproj:
```diff
@@ -47,7 +51,7 @@
     <PackageReference Include="SQLiteNetExtensions.Async" />
     <PackageReference Include="SQLitePCLRaw.bundle_green" />
-	<PackageReference Include="ZXing.Net.Maui.Controls" />
+	  <PackageReference Include="ZXing.Net.Maui.Controls" />
```
Nit pick, please keep indention via spaces.

J0EK3R replied:
fixed that

https://github.com/J0EK3R/brickcontroller2/commit/37d1a5f02fbdd4acf28d7df3941fad99133c20dd

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertiser.cs:
```diff
@@ -0,0 +1,272 @@
+            using (await _asyncLock.LockAsync())
+            {
+                // check if device is connected
+                if (!_connectedDeviceList.Contains(requestingDevice))
```
Nit pick: you can simplify it and call `Remove` here as it returns true / false as well.

J0EK3R replied:
Here it is - but I don't know if I like it; but can live with it ;)

https://github.com/J0EK3R/brickcontroller2/commits/merge/vicocz/StaticDeviceManager/

vicocz commented on BrickController2/BrickController2/DeviceManagement/MK4.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            return address switch
+            {
+                MK6.Device3 => MK4.ChannelStartOffset + 4,
```
Shouldn't it be `MK4`?

J0EK3R replied:
Ohoh, that should not happen! It's fixed...

https://github.com/J0EK3R/brickcontroller2/commit/c152fd57b68fbf94f862879635bd0040e8122a12

vicocz commented on BrickController2/BrickController2/Protocols/MKProtocol.cs:
```diff
@@ -0,0 +1,97 @@
+    /// <param name="ctxValue">ctx value for encryption</param>
+    /// <param name="rfPayload">crypted array</param>
+    /// <returns>size of crypted array</returns>
+    public static int Get_rf_payload(byte[] addr, byte[] data, byte ctxValue, out byte[] rfPayload)
```
Nit pick: I vote for naming it `GetRfPayload`

J0EK3R replied:
...your vote did win ;)

https://github.com/J0EK3R/brickcontroller2/commit/365bd6993ecb3b39aeea1a0e82168191d622b4a2

vicocz commented on BrickController2/BrickController2/Protocols/MKProtocol.cs:
```diff
@@ -0,0 +1,97 @@
+        int addrLength = addr.Length;
+        int dataLength = data.Length;
+
+        byte data_offset = 0x12;    // 0x12 (18)
```
Nit pick: these 2 look as constants.

J0EK3R replied:
You mean addrLength and dataLength?
addrLength at the moment: yes
dataLength no - can be Telegram_Connect (->8) or Telegram_Base (->10)

I am in favor of leaving it as it is... ;)

vicocz commented on BrickController2/BrickController2/Protocols/MKProtocol.cs:
```diff
@@ -0,0 +1,97 @@
+            resultbuf[index + data_offset] = addr[(addrLength - index) - 1];
+        }
+
+        //Buffer.BlockCopy(data, 0, resultbuf, addrLength + data_offset, dataLength);
```
Copying block of bytes is actually good idea, why is it commented out?

J0EK3R replied:
I changed it:

https://github.com/J0EK3R/brickcontroller2/commit/d70db968c05adb7a219edb3af1a21453d43f17a7

vicocz commented on BrickController2/BrickController2/Protocols/MKProtocol.cs:
```diff
@@ -0,0 +1,97 @@
+
+        // calc checksum und copy to array
+        int checksum = CryptTools.Check_crc16(addr, data);
+        resultbuf[result_data_size - 2] = (byte)(checksum & 255);
```
Have you looked at `ByteArrayExtensions` There used to be some set methods, it might be usefull to use/extend them instead of direct access to array.

J0EK3R replied:
Yes, it's done here:

https://github.com/J0EK3R/brickcontroller2/commit/beedb907ad86c2bbfafcb5e6dc62679c0d896fd5

vicocz commented on BrickController2/BrickController2/Protocols/MKProtocol.cs:
```diff
@@ -0,0 +1,97 @@
+
+        int lengthResultArray = addrLength + dataLength + 5;
+
+        if (lengthResultArray > rfPayload.Length)
```
This check should be done at the beggining of this method so as all calculations and reallocations are  not performed.

J0EK3R replied:
I moved the check to the beginning...

https://github.com/J0EK3R/brickcontroller2/commit/beedb907ad86c2bbfafcb5e6dc62679c0d896fd5

vicocz commented on BrickController2/BrickController2/DeviceManagement/MKBaseByte.cs:
```diff
@@ -0,0 +1,164 @@
+        /// <summary>
+        /// after this timespan and all channel's values equal to zero the connect telegram is sent
+        /// </summary>
+        protected TimeSpan _reconnectTimeSpan = TimeSpan.FromSeconds(3);
```
Nit pick: this can be `readonly`

J0EK3R replied:
done

https://github.com/J0EK3R/brickcontroller2/commit/a1b8223fae2e2c663e162d115e6e551a0b145e11

vicocz commented:
> I added a second MK device: MK4
> 
> The special about MK4 is that there is only one advertising telegram for three possible MK4 devices...
> 
> So I had to do some changes: now there is a class `BluetoothAdvertiser `- MK6 devices have one BluetoothAdvertiser instance per device instance - MK4 devices have one BluetoothAdvertiser instance for all three possible device instances...

Sounds reasonable, but I dislike having static fields in MK4 class. Have you considerred e.g. having `BluetoothAdvertiser` registered both as transiend and keyed and pass it via constructor parameter?

vicocz commented on BrickController2/BrickController2/DeviceManagement/DeviceManager.cs:
```diff
@@ -58,12 +58,20 @@ public async Task LoadDevicesAsync()
+                        }
                     }
-                    else
+                    catch
```
Nit pick: if you catch exception, please log it.

J0EK3R replied:
I got an exception when trying to load a corrupt device database.
The database was corrupted because of my testing and changing of branches with different existing devices and their DeviceType enum value not matching any more.
The exception text won't really help...

I vote for leaving it as it is...

vicocz commented on BrickController2/BrickController2/DeviceManagement/DI/DeviceManagementModule.cs:
```diff
@@ -34,6 +38,13 @@ protected override void Load(ContainerBuilder builder)
                     new NamedParameter("settings", settings));
             });
+
+            builder.Register(c => new StaticDeviceFactoryData(DeviceType.MK4, "MK4.0 Device 1", MK4.Device1, Array.Empty<byte>(), [])).As<IStaticDeviceFactoryData>();
```
May be there should be some helper extension method to register these "static" devices so as you do not repeteadly write `StaticDeviceFactoryData`, `IStaticDeviceFactoryData`

J0EK3R replied:
I've added ContainerBuilderExtensions.RegisterDeviceFactory

https://github.com/J0EK3R/brickcontroller2/commit/5d7d1ab62d0ca3c8e8e0d84a82fff7e9a3d82d6d

vicocz commented on BrickController2/BrickController2/DeviceManagement/IStaticDeviceManager.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public interface IStaticDeviceManager : INotifyPropertyChanged
+    {
+        ObservableCollection<IStaticDeviceFactoryData> FactoryDatas { get; }
```
Is it needed to have observable collection here? Can its content be changed in runtime?

J0EK3R replied:
Yes you're right again: the list won't ever change:
I changed ObservableCollection<IStaticDeviceFactoryData> to IEnumerable<IStaticDeviceFactoryData>

https://github.com/J0EK3R/brickcontroller2/commit/2bf3da22130966acf84aec5cc979e2dad592dda3

J0EK3R commented:
I don't understand...?
Can you give me example code?

vicocz commented on BrickController2/BrickController2/DeviceManagement/MKBaseByte.cs:
```diff
@@ -0,0 +1,164 @@
+            _isInitialized = false;
+        }
+
+        public override void SetOutput(int channelNo, float value)
```
Please keep method ordering base on their accessibility public > internal > protected > private.

J0EK3R replied:
OK, done

https://github.com/J0EK3R/brickcontroller2/commit/28a1b67323bcbc2517a2b8b874aabbad4166169e

vicocz commented on BrickController2/BrickController2/Helpers/CryptTools.cs:
```diff
@@ -0,0 +1,153 @@
+﻿namespace BrickController2.Helpers
+{
+    /// <summary>
+    /// static class containing functions needed by encryption algorithm for the advertising data
```
Is it generic or MKs ones?

J0EK3R replied:
Other devices like CADA need that too

vicocz commented on BrickController2/BrickController2/Helpers/CryptTools.cs:
```diff
@@ -0,0 +1,153 @@
+        /// </summary>
+        /// <param name="value">byte to invert</param>
+        /// <returns>inverted byte</returns>
+        public static byte Invert_8(byte value)
```
I would consider make it as extension methods, may be without _8 postfix.

J0EK3R replied:
I'm voting to keep this - I want to keep my implementations (there's another one for Microbit in C) consistent

vicocz replied:
Then I suggest to move it to `Protocols` folder as it make sense it's there to be shared among them.

J0EK3R replied:
Done

https://github.com/J0EK3R/brickcontroller2/commit/29e1a81dc7dd5de6cca4bb2e35139748c19d5dba

vicocz commented on BrickController2/BrickController2/Helpers/CryptTools.cs:
```diff
@@ -0,0 +1,153 @@
+                int currentResult = 0;
+                for (int bitIndex = 0; bitIndex < 8; bitIndex++)
+                {
+                    byte uVar2 = whitening_output(ctx);
```
`whitening_output(ctx)` could be called once (moved outside both cycles) as it process `ctx` only.

J0EK3R replied:
Sorry, you're wrong: `ctx` is `byte[]`
And `WhiteningOutput` is changing that:

```
        private static byte WhiteningOutput(byte[] ctx)
        {
            byte value_3 = ctx[3];
            byte value_6 = ctx[6];
            ctx[3] = ctx[2];
            ctx[2] = ctx[1];
            ctx[1] = ctx[0];
            ctx[0] = ctx[6];
            ctx[6] = ctx[5];
            ctx[5] = ctx[4];
            ctx[4] = (byte)(value_3 ^ value_6);
            return ctx[0];
        }
```

vicocz replied:
Thx, nobody is perfect 😄

vicocz commented on BrickController2/BrickController2/Helpers/CryptTools.cs:
```diff
@@ -0,0 +1,153 @@
+        public static ushort Invert_16(ushort value)
+        {
+            int result = 0;
+            for (int index = 0; index < 0x10; index++)
```
Why don't you use `byte` here for `index` so as you can later skip all the casting and masking to byte.

vicocz replied:
It can be applied in other methods as well.

J0EK3R replied:
Yes, I've fixed that

https://github.com/J0EK3R/brickcontroller2/commit/9d52e555de55058d1f132937243d3eac28a4f590

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,73 @@
+            .SetLegacyMode(true)
+            .SetConnectable(true)
+            .SetScannable(true)
+            .SetInterval(AdvertisingSetParameters.IntervalMedium)
```
`advertisingIterval` parameter should be applied here, I guess.

J0EK3R replied:
Fixed: I added BluetoothLEExtensions to convert BC2 types to Android Bluetooth LE types and applied them

https://github.com/J0EK3R/brickcontroller2/commit/1144fcfe4adb50b3286a28ce6601f9f55278eeb5

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,73 @@
+            .SetConnectable(true)
+            .SetScannable(true)
+            .SetInterval(AdvertisingSetParameters.IntervalMedium)
+            .SetTxPowerLevel(AdvertiseTxPower.Max)
```
`txPowerLevel` parameter should be applied here, I guess.

J0EK3R replied:
Fixed: I added BluetoothLEExtensions to convert BC2 types to Android Bluetooth LE types and applied them

https://github.com/J0EK3R/brickcontroller2/commit/1144fcfe4adb50b3286a28ce6601f9f55278eeb5

vicocz commented on BrickController2/BrickController2/UI/Controls/DeviceChannelSelector.xaml:
```diff
@@ -241,6 +242,77 @@
+            </Grid>
+
+            <!-- MK6 -->
+            <Grid x:Name="MK6Section">
```
Alignment of channel labels does not seem to match position of ports on the image:
![image](https://github.com/user-attachments/assets/71fc9a25-3a61-4897-91d9-e8b577eaa85d)

J0EK3R replied:
let's first clarify whether we take my lousy photos or, for example, those from Amazon

vicocz replied:
Ideally, nice once, well align and sharp, from the vendor. I guess it's not possible for all of them.
So consider taking own pictures or find some reasonable website ;-) (I guess images can be changed if there would be some (legal) confusion.)

J0EK3R replied:
What do you mean with "Alignment of channel labels" and what do you suggest?
Move the labels some pixels vertically right to the box?

J0EK3R replied:
Wasn't it a better solution to show the logical order of the ports from A to F?

vicocz replied:
I suggest to choose an image of the device that allows to place labels (with respect to XAML label capabilities) to reasonable position so as a channel / port can be easily visually checked on real device.

vicocz replied:
Here is example what I meant to align, IMHO label placement does not match the ports on your picture"
![image](https://github.com/user-attachments/assets/75c80de6-370f-4588-8693-f2ab74405cc5)

J0EK3R replied:
Here we are - I did not realize that the "small images" are shown in devicelistpage and the others in the channelselector - I thougt they must show the same but in different sizes...

![grafik](https://github.com/user-attachments/assets/4f217e40-f467-4c1b-90d9-55c3dc21e29c)

![grafik](https://github.com/user-attachments/assets/fda05b19-afff-4210-8aeb-be588aa1b653)

![grafik](https://github.com/user-attachments/assets/971ef515-0e4f-4a80-a334-9944f03d863a)

https://github.com/J0EK3R/brickcontroller2/commit/b49b96b5d80d1a62e22e05c86c99c58feab93497
https://github.com/J0EK3R/brickcontroller2/commit/d7f5d0202536efba128daae6852d483fc243caa5
https://github.com/J0EK3R/brickcontroller2/commit/4b98d7176c972a8d8acdc155202c2f73ad4112ba

vicocz commented on BrickController2/BrickController2/UI/Controls/DeviceChannelSelector.xaml:
```diff
@@ -241,6 +242,77 @@
+                </Grid.ColumnDefinitions>
+
+                <Grid Grid.Column="0">
+                    <controls:ChannelSelectorRadioButton x:Name="MK4Channel0" DeviceType="MK4" Channel="0" HorizontalOptions="End" VerticalOptions="Start"/>
```
Position of channel labels does not match the picture:
![image](https://github.com/user-attachments/assets/83a45837-a8c4-403c-923a-eab24e505e7a)

vicocz replied:
See e.g. image:
![image](https://github.com/user-attachments/assets/534a9950-1190-4b9e-a38c-756100cc0720)

https://www.amazon.de/Mould-King-wiederaufladbarer-Fernbedienung-programmierbar/dp/B09KY8LL3Y

J0EK3R replied:
Same here: let's first clarify whether we take my lousy photos or, for example, those from Amazon

vicocz replied:
In general, take reasonable images, ideally from vendor if possible. But in case of need Amazon might be good as well.
This MK4 picure is probably good enough, just channel alignments should match the reality more.

J0EK3R replied:
OK, I changed the images of MK4 and MK6 and reordered the channels in channelselector.

![grafik](https://github.com/user-attachments/assets/cfec2cb7-d9d7-491a-bc0f-8a64dbbd70e7)

![grafik](https://github.com/user-attachments/assets/a78433a7-84c1-4280-8f5e-0864de940ea6)

https://github.com/J0EK3R/brickcontroller2/commit/a939ebd1615ac0030133ae0b970aa33f6f4f0143

J0EK3R replied:
Here we are - I did not realize that the "small images" are shown in devicelistpage and the others in the channelselector - I thougt they must show the same but in different sizes...

![grafik](https://github.com/user-attachments/assets/4f217e40-f467-4c1b-90d9-55c3dc21e29c)

![grafik](https://github.com/user-attachments/assets/fda05b19-afff-4210-8aeb-be588aa1b653)

![grafik](https://github.com/user-attachments/assets/971ef515-0e4f-4a80-a334-9944f03d863a)

https://github.com/J0EK3R/brickcontroller2/commit/b49b96b5d80d1a62e22e05c86c99c58feab93497
https://github.com/J0EK3R/brickcontroller2/commit/d7f5d0202536efba128daae6852d483fc243caa5
https://github.com/J0EK3R/brickcontroller2/commit/4b98d7176c972a8d8acdc155202c2f73ad4112ba

J0EK3R commented:
First of all - before I start implementing your suggestions - I would like to thank you for your time and work! :)

Commit:
JK: fixed indention via spaces

Commit:
JK: fixed usage of MK6 constant inside MK4, reordered cases

Commit:
JK: renamed Get_rf_payload -> GetRfPayload

Commit:
JK: renamed function, used byte to prevent cast

Commit:
JK: simplified removing

Commit:
JK: added ByteArrayExtensions.SetUInt16 and reworked GetRfPayload

Commit:
JK: using of Buffer.BlockCopy instead of a loop

Commit:
JK: changed ObservableCollection<IStaticDeviceFactoryData> to IEnumerable<IStaticDeviceFactoryData>

Commit:
JK: changed method ordering
public > internal > protected > private.

vicocz commented:
> First of all - before I start implementing your suggestions - I would like to thank you for your time and work! :)

The same for your contributtion. Looking forward to have it working.

vicocz commented on BrickController2/BrickController2/DeviceManagement/MK4.cs:
```diff
@@ -0,0 +1,103 @@
+        private const int ChannelStartOffset = 3;
+
+        /// <summary>
+        /// Telegram wich is sent to connect to MK6.0
```
Nit pick: typo `MK6`

J0EK3R replied:
fixed

https://github.com/J0EK3R/brickcontroller2/commit/cca251c27917e4b8125df7d5953674b1724f1775

vicocz commented on BrickController2/BrickController2/DeviceManagement/MK4.cs:
```diff
@@ -0,0 +1,103 @@
+        /// <summary>
+        /// ManufacturerID for MK
+        /// </summary>
+        public const ushort ManufacturerID = 0xFFF0;
```
This might be part of MK protocol. Or does it change for other MKs blocks?

J0EK3R replied:
I don't know of all MK devices need the same ManufacturerID - but I guess...
Moved it to MKProtocol...

https://github.com/J0EK3R/brickcontroller2/commit/50f26eddbfa9aacb20ea0ccd0958d9aa677eec9d

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertiser.cs:
```diff
@@ -0,0 +1,271 @@
+        /// <summary>
+        /// internal counter increased on data has changed
+        /// </summary>
+        private int _dataVersion = 0;
```
AFAIK accessing `int` might not be an atomic operation, so I guess this should be somehow guarded: e.g. via `volatile` keyword or via `Interlocked` methods, or others.

J0EK3R replied:
Yes, your're right - now, _dataVersion is interlocked.incremented

https://github.com/J0EK3R/brickcontroller2/commit/c37ecc6634b8a07487a755d138b3260b014c5201

Commit:
JK: git-moved MK stuff in folder MouldKing

Commit:
JK: _reconnectTimeSpan is readonly

Commit:
JK: git-moved CryptTools to folder Protocols

Commit:
JK: namespace as File scoped namespace

Commit:
JK: added BluetoothLEExtensions to convert BC2 types to Android Bluetooth LE types

Commit:
JK: typo

Commit:
JK: moved MK ManufacturerID to MKProtocol.ManufacturerID

Commit:
JK: Interlocked.Increment of _dataVersion

Commit:
JK: added ContainerBuilderExtensions.RegisterDeviceFactory

Commit:
JK: reworked editor for static devices

vicocz commented on BrickController2/BrickController2.Android/Extensions/BluetoothLEExtensions.cs:
```diff
@@ -0,0 +1,60 @@
+    /// <returns>Android AdvertisingSetParameters Interval</returns>
+    public static int ToInterval(this AdvertisingInterval advertisingInterval)
+    {
+        switch (advertisingInterval)
```
Nit pick: consider using pattern matching instead, it§s offered as refactoring via VS.
It's IMHO easier to read then.

J0EK3R replied:
OK, if you like - why not

https://github.com/J0EK3R/brickcontroller2/commit/5bb5e41afe0692497ae3569a35cd7c7234964e8d

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertiser.cs:
```diff
@@ -0,0 +1,271 @@
+
+            while (!token.IsCancellationRequested)
+            {
+                currentDataVersion = _dataVersion;
```
You should use `Interlocked.Read` to get it.

J0EK3R replied:
Are you shure?
...I crossed around the [internet](https://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked?view=net-9.0):

> The [Increment](https://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked.increment?view=net-9.0) and [Decrement](https://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked.decrement?view=net-9.0) methods increment or decrement a variable and store the resulting value in a single operation. On most computers, incrementing a variable is not an `atomic operation`, requiring the following steps:
> *   Load a value from an instance variable into a register.
> *  Increment or decrement the value.
> *  Store the value in the instance variable.

...but just getting the value should be `atomic`...

J0EK3R replied:
Interlocked.Read is needed to atomic get a 64 bit value on a 32 bit system...

vicocz commented on BrickController2/BrickController2/DeviceManagement/IStaticDeviceManager.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace BrickController2.DeviceManagement
+{
+    public interface IStaticDeviceManager : INotifyPropertyChanged
```
Is this needed? Will device manager notify about change of `FactoryDataList`? I thought it's static list.

J0EK3R replied:
Yes, you're right - it was a copy-and-paste thing from IDeviceManager - removed it

https://github.com/J0EK3R/brickcontroller2/commit/1560312abd340a1948b6908acb60b4396c290e83

vicocz commented on BrickController2/BrickController2/DeviceManagement/StaticDeviceManager.cs:
```diff
@@ -0,0 +1,20 @@
+    /// This manager stores the needed data to build so called StaticDevices.
+    /// StaticDevices cannot be scanned and so must be inserted manually.
+    /// </summary>
+    internal class StaticDeviceManager : NotifyPropertyChangedSource, IStaticDeviceManager
```
Left comment for `IStaticDeviceManager`, but it relates to the class as well. Why `NotifyPropertyChangedSource` as base when there is nothing to change during runtime?

J0EK3R replied:
Yes, same thing as the interface: copy-and-paste thring from DeviceManager - removed it

https://github.com/J0EK3R/brickcontroller2/commit/65f83edae275b86fcd9c92c858622ee1813c75e0

vicocz commented on BrickController2/BrickController2/Resources/TranslationResources.hu.resx:
```diff
@@ -525,4 +525,10 @@
     <value>Igen</value>
   </data>
+  <data name="AddDevicesManually" xml:space="preserve">
+    <value />
```
May be adding some translation via google translate?

J0EK3R replied:
The ressource editor inserted them - Grrr... I had to look for that...

I removed the empty entries.

https://github.com/J0EK3R/brickcontroller2/commit/9abc316b1ba4cafd533c376bb4211fa78bb2f75b

vicocz commented on BrickController2/BrickController2/Resources/TranslationResources.resx:
```diff
@@ -696,4 +696,13 @@
+    <value>Add devices manually</value>
+  </data>
+  <data name="AddDevicesManually" xml:space="preserve">
+    <value>Add Devices manually</value>
```
How about just description: Manually added devices
at least `Devices` should be `devices` with small letter.

J0EK3R replied:
...I just wanted to "continue the row"

![grafik](https://github.com/user-attachments/assets/df0f8ce6-402d-4088-980d-a67102912e73)

...but I changed it

https://github.com/J0EK3R/brickcontroller2/commit/82ad50a37f2035dc4c6da25e582c57751ba3a9ed

vicocz commented on BrickController2/BrickController2/UI/Controls/DeviceChannelSelector.xaml.cs:
```diff
@@ -55,6 +55,12 @@ public DeviceChannelSelector()
             TechnicMoveChannel4.Command = new SafeCommand(() => SelectedChannel = 6);
             TechnicMoveChannel5.Command = new SafeCommand(() => SelectedChannel = 7);
             TechnicMoveChannel6.Command = new SafeCommand(() => SelectedChannel = 8);
+            MK6Channel0.Command = new SafeCommand(() => SelectedChannel = 0);
```
I miss all 4 `MK4ChannelX.Command assignment.

J0EK3R replied:
Yes, I missed that. :(

...and this is exactly my problem this the current non-generic adding of new devices: you have too many places to edit!

https://github.com/J0EK3R/brickcontroller2/commit/cab303c3a83510eb18e5b93bef479958a8c0ecdb

vicocz commented on BrickController2/BrickController2/UI/ViewModels/DeviceListPageViewModel.cs:
```diff
@@ -36,10 +37,13 @@ public DeviceListPageViewModel(
         public ICommand DeleteDeviceCommand { get; }
         public ICommand DeviceSettingsCommand { get; }
 
+        public bool IsBLEAdvertisingSupported => true;
```
I guess this is temporary, it should be calling `IBluetoothLEService.IsBluetoothLEAdvertisingSupported`

J0EK3R replied:
Yes temporary: to be able to develop in Windows environment

I added a "ToDo" comment

https://github.com/J0EK3R/brickcontroller2/commit/aa41244647e4f76eb732cfd63fe35ef1a7db3305

J0EK3R replied:
now I added a comiler switch for debug mode

https://github.com/J0EK3R/brickcontroller2/commit/5f2262a8bedaf7d7ba5c4046a29e6ae881e65c5c

vicocz commented on BrickController2/BrickController2/UI/ViewModels/StaticDeviceListPageViewModel.cs:
```diff
@@ -0,0 +1,109 @@
+            // get all entries to delete (=> !entry.Selected && entry.ExistingDevice != null)
+            Device[] devicesToDelete = GroupedFactoryDatas.SelectMany(group => group.FindAll(entry => !entry.Selected && entry.ExistingDevice != null).Select(entry => entry.ExistingDevice!)).ToArray();
+
+            if (devicesToCreate.Length > 0)
```
From user perspective I would do both operation under single `ShowProgressDialogAsync` so as it does not blink twice if both added / removed.

J0EK3R replied:
OK - done

https://github.com/J0EK3R/brickcontroller2/commit/a59aeb12725e2b548e37c54529ecd5e0792781b4

vicocz commented on BrickController2/BrickController2/UI/ViewModels/StaticDeviceListPageViewModel.cs:
```diff
@@ -0,0 +1,109 @@
+        private async Task ApplyChangesAsync()
+        {
+            // get all entries to create (=> entry.Selected && entry.ExistingDevice == null)
+            IStaticDeviceFactoryData[] devicesToCreate = GroupedFactoryDatas.SelectMany(group => group.FindAll(entry => entry.Selected && entry.ExistingDevice == null).Select(entry => entry.StaticDeviceFactoryData)).ToArray();
```
These lines are really loooong and hard to read, consider some indention such as (or however you like it):
```
IStaticDeviceFactoryData[] devicesToCreate = GroupedFactoryDatas
    .SelectMany(group => group.FindAll(entry => entry.Selected && entry.ExistingDevice == null)
            .Select(entry => entry.StaticDeviceFactoryData))
    .ToArray();
```

J0EK3R replied:
Hey, you should mark that as **nitpick** ;)

I changed it:

https://github.com/J0EK3R/brickcontroller2/commit/c51e21958d615025e9e816c503307b7e425542da

vicocz commented on BrickController2/BrickController2/UI/ViewModels/StaticDeviceListPageViewModel.cs:
```diff
@@ -0,0 +1,109 @@
+
+            // translate list to dict (key is DeviceType)
+            var dict = staticDeviceManager.FactoryDataList.GroupBy(o => o.DeviceType)
+                       .ToDictionary(g => g.Key, g => g.Select(x => new StaticDeviceEntry(x, GetDeviceInstance(x))).ToList());
```
It seems to me you do not need to transform it to dictionary, just use grouping and transformation of elements as you do in dictionary creation.
Additionally evaluation via `ToList` is redundant as it's evaluated one miliseconds later on the next line.

J0EK3R replied:
Hmm... I can't find a better way - can you give me your suggestion as example, please?

vicocz replied:
E.g.:
```
            var groups = staticDeviceManager.FactoryDataList
                .GroupBy(o => o.DeviceType, x => new StaticDeviceEntry(x, GetDeviceInstance(x)));

            GroupedFactoryDatas.AddRange(groups.Select(item => new StaticDeviceGroup(item.Key, item.ToList())));
```

J0EK3R replied:
OK, thank you :)

I'm learning so much... :)

https://github.com/J0EK3R/brickcontroller2/commits/merge/vicocz/StaticDeviceManager/

Commit:
JK: replaces classicla switch-case by pattern mathcing

Commit:
JK: removed INotifyPropertyChanged

Commit:
JK: StaticDeviceManager not based on NotifyPropertyChangedSource any more

Commit:
JK: removed empty entries

Commit:
JK: changed "Add Devices manually" to "Manually add Devices"

Commit:
JK: added ToDo comment

Commit:
JK: improved readability by indention

Commit:
JK: improved readability by indention

Commit:
JK: just one Progress Dialog to Add/Remove sevices

Commit:
JK: bugfix - missing Command assignment

Commit:
JK: improved styling

Commit:
JK: in debug mode DeviceListPageViewModel.IsBLEAdvertisingSupported = true

Commit:
JK: revised BluetoothAdvertiser
* usage of AutoResetEvent to signal changes of values immediately
-> bugfix: connect telegram is sent after _reconnectTimeSpan elepsed

J0EK3R commented:
I revised the BluetoothAdvertsing part:
There was a bug sending the connect telegram if all channels are zero and the reconnect timespan elapsed.

And to react quickly on value changes I'm using a AutoResetEvent.
So ProcessOutputs is not needed to be async any more.

How do you like it?
...I'm expecting the worst! ;)

https://github.com/J0EK3R/brickcontroller2/commit/e92662d174e39bfe2a67535e4c1fdb76d410b861

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertiser.cs:
```diff
@@ -241,30 +268,18 @@ private async Task StopOutputTaskInternalAsync()
                 }
 
-                await Task.Delay(_cyclicLoopWaitTimeSpan, token).ConfigureAwait(false);
+                Thread.Sleep(1); // prevent loop without sleep
```
I suggest to keep `await Task.Delay(1)` and leave the method async.

J0EK3R replied:
Can you tell me why?
The method ProcessOutputs and the loop inside is run as Task - what's the advantage doing the things async?
(I'm an old threading-guy just starting with the async business)

J0EK3R replied:
OK, I think I got it...

I revised BluetoothAdvertsing
* ProcessOutputs is async again
* UpdateAdvertisedData only if data changes

https://github.com/J0EK3R/brickcontroller2/commit/b2240c517b6b3f3b8ceec2ada928fe27163fb216

vicocz commented:
> I revised the BluetoothAdvertsing part: There was a bug sending the connect telegram if all channels are zero and the reconnect timespan elapsed.
> 
> And to react quickly on value changes I'm using a AutoResetEvent. So ProcessOutputs is not needed to be async any more.
> 
> How do you like it? ...I'm expecting the worst! ;)
> 
> [J0EK3R@e92662d](https://github.com/J0EK3R/brickcontroller2/commit/e92662d174e39bfe2a67535e4c1fdb76d410b861)

I think signal sending is OK, may be there might be some design changes in TryGetTelegram, need to have a look at it in VS....

Commit:
JK: revised BluetoothAdvertsing
* ProcessOutputs -> ProcessOutputsAsync
* UpdateAdvertisedData only if data changes

vicocz commented:
Approved, just to let CI build to run. please wait with the merge yet.

vicocz commented on BrickController2/BrickController2/UI/Pages/StaticDeviceListPage.xaml:
```diff
@@ -0,0 +1,76 @@
+                                        <ColumnDefinition Width="Auto" />
+                                        <ColumnDefinition Width="*" />
+                                    </Grid.ColumnDefinitions>
+                                    <Label Grid.Column="1" Text="{Binding StaticDeviceFactoryData.Address}" FontSize="Micro" VerticalTextAlignment="Center" />
```
May be this should have larger font, this one seems to be quite small....
![image](https://github.com/user-attachments/assets/0b421771-92f0-42a8-b7a8-a1602738e9ac)

J0EK3R replied:
Yeah, looks much better now (FontSize Micro -> Medium)

https://github.com/J0EK3R/brickcontroller2/commit/a5cb1c272be205ac08df76d012ac31233408a758

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    // JK: commented out because app crashes
+    //public override void OnAdvertisingSetStopped(AdvertisingSet? advertisingSet)
```
I was curious, uncommented and nothing happend, may be because I have no MK device.
Could you recheck it and share some expcetion, call stack if it craches?

J0EK3R replied:
It does not matter if you physically have a MK device -> it's a one-way communication...

Really curious - whily testing, I did not get any exception...?

I removed the comments and it's used again to set _advertisingSet to null.

https://github.com/J0EK3R/brickcontroller2/commit/f457fe0638a22ec6c54f6ea377f904ce54f73318

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,74 @@
+
+    public void UpdateAdvertisedData(ushort manufacturerId, byte[] rawData)
+    {
+        if (_advertisingSet != null)
```
One interesting thing when I tried to debug it locally, `UpdateAdvertisedData` gets called before `OnAdvertisingSetStarted` and it means `_advertisingSet` is null.
I guess connect datagram is sent as part of StartAdvertise, so probably not something serious. May we we can add some reasonable delay to `StartOutputTaskInternal` before `ProcessOutputsAsync` is started.

J0EK3R replied:
OK - suggestion:
let's make the methods in IBluetoothLEAdvertiserDevice async and await them...

How do you like that

https://github.com/J0EK3R/brickcontroller2/commit/d313953dc4858262214ade4d9f14d490897b74a2

vicocz replied:
Can be, just adjust that waiting (apply some timeout).

Commit:
JK: simplyfied creation of GroupedFactoryDatas

Commit:
JK: increased FontSize from Micro ro Medium

Commit:
JK: typo and unification
* "Manually add Devices" -> "Manually add devices"
* "Add devices manually" -> "Manually add devices"

Commit:
JK: using OnAdvertisingSetStopped again

Commit:
JK: IBluetoothLEAdvertiserDevice with async methods

J0EK3R commented:
My suggestion: `ManualDevice` was miles better naming than `StaticDevice`
...I could change all the StaticDevice naming to ManualDevice?

What do you think?

Commit:
JK: fixed async name TryDisconnect -> TryDisconnectAsync

Commit:
JK: Dispose _bleAdvertiserDevice
* renamed GetBluetoothLEAdvertiserDevice -> CreateBluetoothLEAdvertiserDevice
This change clarifies the need to dispose the _bleAdvertiserDevice

Commit:
JK: don't await TaskCompletionSource on any exception

Commit:
JK: on exception set TaskCompletionSource member to null

vicocz commented:
> My suggestion: `ManualDevice` was miles better naming than `StaticDevice` ...I could change all the StaticDevice naming to ManualDevice?
> 
> What do you think?

Why not, slightly better name.

Commit:
JK: renamed StaticDevice -> ManualDevice

J0EK3R commented:
I renamed StaticDevice types and files and stuff to ManualDevice.
It's much better... :)

https://github.com/J0EK3R/brickcontroller2/commit/14142c200f38392d45686bc3c9afbf2f87970b7f

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,134 @@
+                    null,
+                    this);
+
+                await advertisingStarted.Task;
```
How about having some limited wait here in case `OnAdvertisingDataSet` is not called. Could you check it please?
The same for `StopAdvertiseAsync` as well.

J0EK3R replied:
Yes, good idea - here it is:

https://github.com/J0EK3R/brickcontroller2/commit/b8f5fedb454f6b64bdf4268658d0d00a3bc6c272

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,134 @@
+            {
+                // https://developer.android.com/reference/android/bluetooth/le/BluetoothLeAdvertiser#startAdvertisingSet(android.bluetooth.le.AdvertisingSetParameters,%20android.bluetooth.le.AdvertiseData,%20android.bluetooth.le.AdvertiseData,%20android.bluetooth.le.PeriodicAdvertisingParameters,%20android.bluetooth.le.AdvertiseData,%20android.bluetooth.le.AdvertisingSetCallback)
+                // possible exception: IllegalArgumentException
+                _advertiser?.StartAdvertisingSet(
```
`_advertiser` is always not null, therfore `?.` is useless here.

J0EK3R replied:
Yes - it's useless and removed now...

https://github.com/J0EK3R/brickcontroller2/commit/3626e5311dec23cc7fb447198b45c36a78108b02

vicocz commented on BrickController2/BrickController2.Android/PlatformServices/BluetoothLE/BluetoothLEAdvertiserDevice.cs:
```diff
@@ -0,0 +1,134 @@
+
+            try
+            {
+                _advertiser?.StopAdvertisingSet(this);
```
_advertiser is always not null, therfore `?.` is useless here.

J0EK3R replied:
Same here: useless und removed

https://github.com/J0EK3R/brickcontroller2/commit/3626e5311dec23cc7fb447198b45c36a78108b02

Commit:
JK: removed useless ?-operator

Commit:
JK: awaiting TaskCompletionSource is set or WaitAsyncTimeout expires

vicocz commented:
Please list MK4 - 6 in README as well.

Commit:
JK: added MK4.0 and MK6.0 to README.md

J0EK3R commented:
I added both devices to README.md
* Mould King 4.0 Powered Module
* Mould King 6.0 Powered Module

https://github.com/J0EK3R/brickcontroller2/commit/13bc3b015edec89f6ef7d11aba71cfe5d750d2a2

Commit:
JK: changed images for MK4 and MK6 and order of channels in channelselector

Commit:
JK: changed MK6 image for channelselector

Commit:
JK: changed MK4 image for devicepage

Commit:
JK: optimized image WidthRequest and HeightRequest

J0EK3R commented:
How would you like the devices in devicelistpage ordered by name then address?

I don't like the current ordering...

![grafik](https://github.com/user-attachments/assets/b82b2cce-a320-4017-b6f9-9a6f9d82f89f)

vicocz commented:
> How would you like the devices in devicelistpage ordered by name then address?
> 
> I don't like the current ordering...
> 
I have no strong opinion related to sorting, but on the other hand there might be people who do not like alphabetical sorting as well. In past I noted the idea to enable drag&drop ordering elsewhere (https://github.com/vicocz/brickcontroller2/issues/25).

May be there might be some features to sort and filter all the lists. Hard to say what's supported in MAUI and common in mobile apps world.

Anyway, please not in this PR as it's big enough and it takes me enormous amount of time to review it.

vicocz commented:
@J0EK3R  I guess this seems to be mergeble PR right now, could not test it by myself, so waiting for your signal to merge it.

J0EK3R commented:
Please merge it :)

J0EK3R commented:
I want to thank you again for your time and work!
I know how much time all your reviews and nitpicks have taken up.

I'm planning to include the rest of my BLE Advertising devices...
...or do you need some air to breathe? ;)

I think the following PR will pass faster...

vicocz commented:
> I want to thank you again for your time and work! I know how much time all your reviews and nitpicks have taken up.
> 
> I'm planning to include the rest of my BLE Advertising devices... ...or do you need some air to breathe? ;)
> 
> I think the following PR will pass faster...

It depends...

If it's obvious bug fix, do not hesitate and create PR for that.
If it's adding of new device which does not include huge changes, do the same, just leave it as regular PR if it's complete, working and ready to review. Otherwise create it as DRAFT and keep working on that.
In general, larger PR, the longer it takes.
